packages/react-reconciler/src/ReactFiberHooks.new.js

```diff
@@ -1012,7 +1012,8 @@ function useMutableSource<Source, Snapshot>(
       const latestSetSnapshot = refs.setSnapshot;
 
       try {
-        latestSetSnapshot(latestGetSnapshot(source._source));
+        const value = latestGetSnapshot(source._source);
+        latestSetSnapshot(() => value);
 
         // Record a pending mutable source update with the same expiration time.
         const suspenseConfig = requestCurrentSuspenseConfig();
```

Comment:
(nb: we still need some wrapper for the exceptions a few lines below this – we could maybe make something different for that to avoid allocating the function in the success case, but… I feel like this is fine?)